tests/unit/s2n_ktls_io_test.c

```diff
@@ -286,5 +300,217 @@ int main(int argc, char **argv)
+
+        /* Handle a S2N_ERR_IO error */
+        {
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            struct s2n_test_ktls_io_fail_ctx io_ctx = {
+                .errno_code = EINVAL,
+                .return_value = -1,
+            };
+            EXPECT_OK(s2n_ktls_set_recvmsg_cb(client, s2n_test_ktls_recvmsg_fail, &io_ctx));
+
+            uint8_t recv_buf[S2N_TLS_MAXIMUM_FRAGMENT_LENGTH] = { 0 };
+            s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+            uint8_t recv_record_type = 0;
+            size_t bytes_read = 0;
+            EXPECT_ERROR_WITH_ERRNO(
+                    s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read),
+                    S2N_ERR_IO);
+            /* Blocked status intentionally not reset to preserve legacy s2n_send behavior */
+            EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+
+            EXPECT_EQUAL(io_ctx.invoked_count, 1);
+        };
+
+        /* Simulate EOF and handle a S2N_ERR_CLOSED error */
+        {
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            ssize_t return_value_eof = 0;
+            struct s2n_test_ktls_io_fail_ctx io_ctx = {
+                .return_value = return_value_eof,
+            };
+            EXPECT_OK(s2n_ktls_set_recvmsg_cb(client, s2n_test_ktls_recvmsg_fail, &io_ctx));
+
+            uint8_t recv_buf[S2N_TLS_MAXIMUM_FRAGMENT_LENGTH] = { 0 };
+            s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+            uint8_t recv_record_type = 0;
+            size_t bytes_read = 0;
+            EXPECT_ERROR_WITH_ERRNO(
+                    s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read),
+                    S2N_ERR_CLOSED);
+            /* Blocked status intentionally not reset to preserve legacy s2n_send behavior */
+            EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+
+            EXPECT_EQUAL(io_ctx.invoked_count, 1);
+        };
+
+        /* Simulate control message truncated via MSG_CTRUNC flag and handle a S2N_ERR_KTLS_BAD_CMSG error */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_ktls_io_stuffer_pair io_pair = { 0 },
+                    s2n_ktls_io_stuffer_pair_free);
+            EXPECT_OK(s2n_test_init_ktls_io_stuffer(server, client, &io_pair));
+
+            /* The stuffer mock IO is used to ensure `cmsghdr` is otherwise properly constructed
+             * and that the failure occurs due to the MSG_CTRUNC flag. */
+            io_pair.client_in.recv_msg_flags = MSG_CTRUNC;
```

Comment:
You only use this in one place, so I would not suggest including it in the standard stuffer implementation.